mtags/src/main/scala/scala/meta/internal/metals/ReportsProvider.scala

```diff
@@ -0,0 +1,121 @@
+package scala.meta.internal.metals
+
+import java.io.File
+import java.nio.file.Files
+import java.nio.file.Path
+
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.io.AbsolutePath
+import scala.meta.io.RelativePath
+
+class Reports(workspace: AbsolutePath) {
+  lazy val reportsDir: AbsolutePath = {
+    val dir = workspace.resolve(Reports.reportsDir)
+    Files.createDirectories(dir.toNIO)
+    dir
+  }
+
+  val unsanitized =
+    new ReportsProvider(workspace, Reports.reportsDir.resolve("metals-full"))
+  val incognito =
+    new ReportsProvider(workspace, Reports.reportsDir.resolve("metals"))
+  val bloop =
+    new ReportsProvider(workspace, Reports.reportsDir.resolve("bloop"))
+
+  def all: List[ReportsProvider] = List(unsanitized, incognito, bloop)
+  def allToZip: List[ReportsProvider] = List(incognito, bloop)
+
+  def cleanUpOldReports(
+      maxReportsNumber: Int = Reports.MAX_NUMBER_OF_REPORTS
+  ): Unit = {
+    all.foreach(_.cleanUpOldReports(maxReportsNumber))
+  }
+
+  def deleteAll(): Unit = {
+    all.foreach(_.deleteAll())
+    Files.delete(reportsDir.resolve(Reports.ZIP_FILE_NAME).toNIO)
+  }
+}
+
+class ReportsProvider(workspace: AbsolutePath, pathToReports: RelativePath) {
+  private lazy val reportsDir = {
+    val dir = workspace.resolve(pathToReports)
+    Files.createDirectories(dir.toNIO)
+    dir
+  }
+
+  private lazy val userHome = Option(System.getProperty("user.home"))
+
+  def createReport(name: String, text: String): AbsolutePath = {
+    val path = reportsDir.resolve(s"r_${name}_${System.currentTimeMillis()}")
+    path.writeText(sanitize(text))
+    path
+  }
+
+  def createReport(name: String, text: String, e: Throwable): AbsolutePath =
+    createReport(
+      name,
+      s"""|$text
+          |Error message: ${e.getMessage()}
+          |Error: $e
+          |""".stripMargin
+    )
+
+  private def sanitize(text: String) = {
+    val textAfterWokspaceReplace =
+      text.replace(workspace.toString(), Reports.WORKSPACE_STR)
+    userHome
+      .map(textAfterWokspaceReplace.replace(_, Reports.HOME_STR))
+      .getOrElse(textAfterWokspaceReplace)
+  }
+
+  def cleanUpOldReports(
+      maxReportsNumber: Int = Reports.MAX_NUMBER_OF_REPORTS
+  ): List[Report] = {
+    val reports = getReports
+    if (reports.length > maxReportsNumber) {
+      val filesToDelete = reports
+        .sortBy(_.timestamp)
+        .slice(0, reports.length - maxReportsNumber)
+      filesToDelete.foreach { f => Files.delete(f.toPath) }
+      filesToDelete
+    } else List()
+  }
+
+  def getReports: List[Report] = {
```

Comment:
```suggestion
  def getReports(): List[Report] = {
```
just my OCD since it's a side effect function(reads from IO?)